### ISSUE TITLE
Add support for authenticator choice

### ISSUE THREAD

**Issue by sdsantos** [hirosystems/connect#101]
The new alpha release of this library does not support mobile authentication. It always leads users to install the Stacks Wallet browser extension, which is only possible for users using desktop Chrome and Firefox. This locks out a huge part of the audience, using mobile devices, or other browsers.

Here are some alternative proposals:

1) We are developing an authenticator for Android ([Circles](https://github.com/blocoio/stacks-circles-app)), and I believe one for iOS is also being built by the community. The library could detect that a mobile device is being used, and lead users to a native alternative. The downside is that these alternatives would be hardcoded in the library

2) The auth popup could lead users to a well-known url. Authenticators could intercept that URL to proceed to sign-in/sign-up flow, if they are installed in the users device. And that url would fallback to a website pointing users to the authenticators they could install for their device.

Maybe this was already discussed by the Hiro team, but I couldn't find a public discussion on the matter. And while this isn't solved, it puts new apps that need transaction support in the tough spot of leaving out their whole mobile audience. And from our perspective, we would be releasing Circles with support for current apps, but it wouldn't support the next release of this library.

**Comment by jasperjansz** [comment]
Here's the first draft of the flow: https://www.figma.com/file/ksq6tnGDWcm6icDNShdENt/Connect?node-id=11%3A2447. Would this work for you guys — or do you have another approach in mind?

**Comment by yknl** [comment]
> Here's the first draft of the flow: https://www.figma.com/file/ksq6tnGDWcm6icDNShdENt/Connect?node-id=11%3A2447. Would this work for you guys — or do you have another approach in mind?

Thanks for putting this together, I left some comments on the Figma file.

Maybe I'm misinterpreting the purpose of this Figma flow. I was more looking for a diagram that outlines what happens behind the scenes with wallet connect. 

For example:
- How information will be passed from the app to the wallet
- How does app connect remember that the user prefers mobile wallet vs browser extension
- Which protocols will be used in different scenarios, like desktop browser vs mobile browser or native apps

**Comment by models-72** [comment]
Just had a meeting with Mark about the stacks authenticator.  I am interested in learning about the specification that will be produced by Ken and would like to attend any meetings where we discuss this in the future.  If I could understand the specification well enough I would like to try to build an authenticator from scratch in a non-javascript language.

**Comment by markmhx** [comment]
Some context that wasn't captured here earlier: 

@yknl is going to help out by producing the technical diagram that [he inquired about previously](https://github.com/blockstack/connect/issues/101#issuecomment-840647887). Once he has that ready, we can all meet for another community hangout on Discord to review that diagram alongside the UI designs produced by @jasperjansz.

I'm hoping we can schedule this meeting in the second half of June to keep momentum going and decide next steps from there.

**Comment by yknl** [comment]
As @markmhx mentioned, Secret Key Labs is working on a proof-of-concept for mobile transaction signing support using wallet connect 2.0 and Xverse wallet. We'll publish the details relatively soon.

**Comment by Filmaluco** [comment]
@yknl any progress?

**Comment by yknl** [comment]
> @yknl any progress?

We're working on it. Is there a specific flow or use case you are looking for?

**Comment by Filmaluco** [comment]
@yknl yes. The Stack Circles Android Since intercepts app.blockstack.org to authenticate apps on Android devices. Using the latest version of the @stack/connect library, users are prompted to use the Hiro Extension, and never go through app.blockstack.org.
I need to use a previous version of the library, and but can't look for it since the library has no changelog and was rebranded from the old @blockstack/connect.
So, basically, we need a previous version of connect compatible with the latest stacks libraries. Or a new protocol to authenticate on mobile devices.

**Comment by markmhx** [comment]
@Filmaluco Presumably you want "a new protocol to authenticate on mobile devices." more than "a previous version of connect compatible with the latest stacks libraries", correct?

Since you want mobile Android users to authenticate with Circles, _not_ with Blockstack Connect at app.blockstack.org.

**Comment by Filmaluco** [comment]
@markmhx With no deadline to a new protocol we need an alternative for our projects, this issue was open on Feb and we still have no timeline.

**Comment by yknl** [comment]
@Filmaluco I think what you're looking for is a mobile web to mobile authenticator link correct? We're working on multiple protocols that link apps from web, mobile web, native mobile to mobile and desktop wallets. Currently focused on the web to mobile wallet link using wallet connect 2.0. The proposal for a mobile web to native mobile wallet link was through app links. If you're blocked on this and would like to contribute we should chat.

**Comment by markmhx** [comment]
> this issue was open on Feb and we still have no timeline.

I agree and I'd love to see more measurable progress towards mobile support in particular.

@yknl @Filmaluco perhaps the two of you could sync up and figure out how to combine efforts towards setting such a timeline?

Given all the other work on the UserX team's plate at Hiro these days, it's not realistic that we'll be able to own this project on our end for the coming months. As such, it's imperative that ecosystem members who need this functionality step in and push this forward if we all want it available by EOY.

**Comment by Filmaluco** [comment]
@markmhx at the moment we don't have time and resources to spend defining the protocol and in evangelization. It's not realistic that we'll be able to own this project either. 

As of pushing this functionality forward, we wouldn't need to if the app.blockstack.org was not being deprecated or a **equal** alternative was presented. 

This is not just a developers problem, mobile users can't authenticate on the latest stacks ecosystem vision, an entire group of users will be discarded with these changes, changes we have no control and are currently centralized.

**Comment by markmhx** [comment]
> we wouldn't need to if the app.blockstack.org was not being deprecated 

To be clear, we won't sunset app.blockstack.org until a mobile alternative is available: https://github.com/blockstack/stacks-wallet-web/issues/970

Older versions of Connect do indeed hand off to app.blockstack.org whereas the newest version hands off to the Hiro Wallet extension. I imagine it should be relatively straightforward to submit a PR to Connect that allows the developer to pass a parameter (`useLegacyAuth` or the like) that sends the user to app.blockstack.org on click? Perhaps by borrowing old code from a previous version.

**Comment by Filmaluco** [comment]
@markmhx that's a good idea, but for Mobile/Android developers such as our self's the best we can do is open issues... changing such complex code in a different environment is not something we are comfortable doing.

**Comment by markmhx** [comment]
Hey all, I was talking with @dantrevino about this the other day and it seems we're due for a synchronous ecosystem meeting to discuss the state of authenticator choice (and mobile auth in particular) as we enter Q4 2021.

We haven't had the capacity at Hiro to push this as long as much I would have liked in 2021 so far. And while I'm aware that @yknl at Xverse is cooking up a near-term solution for iOS and Android, I'd love to get a better sense as a group as to our most pressing needs and capabilities.

If you're interested in joining, please upvote this comment with an emoji. I'll organize it for Discord in a couple of weeks from now, reaching out to whoever votes to coordinate a time that works for all.

**Comment by friedger** [comment]
Maybe authenticator choice can be a solution for https://github.com/blockstack/stacks-wallet-web/issues/1412 : One wallet per derivation path.

**Comment by markmhx** [comment]
https://github.com/hirosystems/connect/pull/269 has been submitted for displaying hardcoded options for both Hiro Wallet and Xverse

**Comment by janniks** [comment]
_closing this issue for now, but there will be more connect interface work soon_
